pgx/src/trigger_support/pg_trigger.rs

```diff
@@ -159,51 +193,39 @@ impl PgTrigger {
             Ok(None)
         }
     }
+
     /// The `PgRelation` corresponding to the trigger.
     ///
     /// # Panics
     ///
     /// If the relation was recently deleted, this function will panic.
-    ///
-    /// # Safety
-    ///
-    /// The caller should already have at least AccessShareLock on the relation ID, else there are nasty race conditions.
-    ///
-    /// As such, this function is unsafe as we cannot guarantee that this requirement is true.
-    pub unsafe fn relation(&self) -> Result<crate::PgRelation, PgTriggerError> {
-        let relation = PgRelation::open(self.relation_data.rd_id);
-        Ok(relation)
+    pub fn relation(&self) -> Result<crate::PgRelation, PgTriggerError> {
+        // SAFETY:  the creator of this PgTrigger has asserted that the provided "fcinfo" struct
+        // contains proper trigger data, and that would include the fact that the relation already
+        // has at least a share lock created by Postgres
```

Comment:
Have we checked the fcinfo struct for this at this point? Can we? It's one thing to rely on Postgres getting the layout correct, but I don't like relying on Postgres not null-initializing a field.

Reply:
I don't know that we can.  Especially the invariant of "table should have an AccessShareLock".

I'll take a look at the `from_fcinfo()` function and see if there's a little bit more validation we can do.

Reply:
Is it possible for someone to wrap this struct in an `Arc`, hang it in a `static` like `static SHENANIGANS: Arc<Mutex<Arc<PgTrigger>` and then later access this struct and attempt to call this function on it when we don't have a lock?

Reply:
That's the least of the problems:

```rust
pub struct PgTrigger {
    trigger: pg_sys::Trigger,
    trigger_data: PgBox<pgx_pg_sys::TriggerData>,
    relation_data: pg_sys::RelationData,
    #[allow(dead_code)]
    fcinfo: pg_sys::FunctionCallInfo,
}
```

We can maybe add a lifetime to `PgTrigger` and tie it to the `fcinfo` used to create it?  That would prevent such things, yeah?

Reply:
Hm. I don't know if a lifetime is enough. The appropriate interface may be to accept a closure and run that with "safe access to the unsafe accessors", without yielding that type to the outside world. I'm not sure, I haven't seen many triggers so I don't know what they look like.

Reply:
```rust
#[pg_trigger]
fn trigger_example(
    trigger: &pgx::PgTrigger,
) -> Result<PgHeapTuple<'_, impl WhoAllocated>, TriggerError> {
    let old = trigger.current().ok_or(TriggerError::NullOld)?;

    let mut current = old.into_owned();
    let col_name = "title";

    if current.get_by_name(col_name)? == Some("Fox") {
        current.set_by_name(col_name, "Bear")?;
    }

    Ok(current)
}
```

That's an example from our, um, examples.

Reply:
Postgres is going to give us the underlying `fcinfo` struct properly setup.

If the concern is if a user can then hold onto this `trigger: &pgx::Trigger` reference for longer than this function... I'm all ears on how to fix that.  I guess by virtue of the fact it is a reference already means it has a lifetime associated with it.

Reply:
Conditionalizing it on `&mut pg_sys::FunctionCallInfo` might be enough since that's exclusive access, however, if a programmer gets access to that `&mut pg_sys::FunctionCallInfo`, they can run `mem::take` on the FunctionCallInfo. That would be invalid since it's not `Default`, but...

```rust
mem::replace(&mut fcinfo, ptr::null_mut())
```
is not.

It may need `Pin` if they can touch it, assuming we want to be able to rely on it being valid after its use. The problem with it currently is that the current unsafe constructor (correctly) consumes the fcinfo.

Reply:
Well that `&PgTrigger` is fine.

Reply:
Okay, if `&mut PgTrigger` or owned PgTrigger is only accessible from inside the "function wrapper" and not the "user function" then we're cool.

Reply:
I believe that https://github.com/tcdi/pgx/commit/5fc5a19f78e617adc494ea42aedc8ad13bdb6db4 addresses this nicely.  make sure to read its commit message  ;)